zerver/lib/message.py

```diff
@@ -942,6 +942,38 @@ def bulk_access_messages(
     return filtered_messages
 
 
+def bulk_access_stream_messages_query(
+    user_profile: UserProfile, messages: QuerySet[Message], stream: Stream
+) -> QuerySet[Message]:
+    """This function mirrors bulk_access_messages, above, but applies the
+    limits to a QuerySet and returns a new QuerySet which only
+    contains messages in the given stream which the user can access.
+    Note that this only works with streams.  It may return an empty
+    QuerySet if the user has access to no messages (for instance, for
+    a private stream which the user is not subscribed to).
+
+    """
+
+    messages = messages.filter(realm_id=user_profile.realm_id, recipient_id=stream.recipient_id)
+
+    if stream.is_public() and user_profile.can_access_public_streams():
+        return messages
```

Comment:
Can this use `can_access_stream_history`? That handles the guest corner cases slightly better, and also it's nice to avoid as many additional copies of security logic as possible.

Reply:
Yeah I think it needs a bit of a refactor or other treatment where you optionally pass in some previously fetched data about whether someone's subscribed. 

I guess that should just be part of https://github.com/zulip/zulip/pull/21970 in any case.